django/contrib/messages/storage/cookie.py

```diff
@@ -122,26 +143,57 @@ def _store(self, messages, response, remove_oldest=True, *args, **kwargs):
         returned), and add the not_finished sentinel value to indicate as much.
         """
         unstored_messages = []
-        encoded_data = self._encode(messages)
+
+        serialized_messages = MessagePartSerializer().dumps(messages)
+        encoded_data = self._encode_parts(serialized_messages)
         if self.max_cookie_size:
             # data is going to be stored eventually by SimpleCookie, which
             # adds its own overhead, which we must account for.
             cookie = SimpleCookie()  # create outside the loop
 
-            def stored_length(val):
-                return len(cookie.value_encode(val)[1])
+            def is_too_large_for_cookie(data):
+                return bool(
+                    data and len(cookie.value_encode(data)[1]) > self.max_cookie_size
+                )
+
+            def compute_msg(some_serialized_msg):
+                return self._encode_parts(
+                    some_serialized_msg + [json.dumps(self.not_finished)],
+                    encode_empty=True,
+                )
 
-            while encoded_data and stored_length(encoded_data) > self.max_cookie_size:
+            if is_too_large_for_cookie(encoded_data):
                 if remove_oldest:
-                    unstored_messages.append(messages.pop(0))
+                    idx = bisect_keep_right(
+                        serialized_messages,
+                        fn=lambda m: is_too_large_for_cookie(compute_msg(m)),
+                    )
+                    unstored_messages = messages[:idx]
+                    encoded_data = compute_msg(serialized_messages[idx:])
                 else:
-                    unstored_messages.insert(0, messages.pop())
-                encoded_data = self._encode(
-                    messages + [self.not_finished], encode_empty=unstored_messages
-                )
+                    idx = bisect_keep_left(
+                        serialized_messages,
+                        fn=lambda m: is_too_large_for_cookie(compute_msg(m)),
+                    )
+                    unstored_messages = messages[idx:]
+                    encoded_data = compute_msg(serialized_messages[:idx])
+
         self._update_cookie(encoded_data, response)
         return unstored_messages
 
+    def _encode_parts(self, messages, encode_empty=False):
+        """
+        Return an encoded version of the serialized messages list which can be stored as
+        plain text.
+
+        Since the data will be retrieved from the client-side, the encoded data
+        also contains a hash to ensure that the data was not tampered with.
+        """
+        if messages or encode_empty:
+            return self.signer.sign_object(
+                messages, serializer=MessagePartGatherSerializer, compress=True
+            )
+
     def _encode(self, messages, encode_empty=False):
```

Comment:
This is no longer used and could be removed, or am I missing something?

Reply:
Indeed, it's unused, at least only used in tests.

There are features that the current `MessagePartSerializer` + `MessagePartGatherSerializer` classes don't support: for instance passing a simple `Message` object, and then decoding it (since we do a part of the encoding manually, and it currently handles list of `Message`s).
I guess it would make sense to have classes that have the same features. Do you agree? :)

Reply:
`MessageSerializer` is used only for deserializing in the code, while it still has `dumps` method, which is used from `MessagePartSerializer` instead. Similar is with `CookieStorage._encode_parts` and `CookieStorage._encode` – they are nearly same, just use different serializer (and take different thing as `messages`).

Having both `MessageSerializer.dumps`/`MessagePartSerializer.dumps` and `CookieStorage._encode_parts`/`CookieStorage._encode` while they essentially reimplement the same logic might cause maintenance issues.

If keeping `CookieStorage._encode` is desired, it IMHO should be using part serializer to reuse the code, something like:

```
    return self._encode_parts(MessagePartSerializer().dumps(messages), encode_empty)
```

The `MessageSerializer.dumps` can be then removed as it is no longer used.